vault/src/execution.rs

```diff
@@ -270,9 +271,108 @@ impl Request {
-            let tx_metadata = btc_rpc.wait_for_transaction_metadata(txid, num_confirmations).await?;
+            tracing::info!("Awaiting bitcoin confirmations for {txid}");
+
+            let txid_copy = txid.clone();
+
+            let fee_rate_subscription = parachain_rpc.on_fee_rate_change();
+            let fee_rate_subscription = BroadcastStream::new(fee_rate_subscription);
+            let subscription = fee_rate_subscription
+                .map_err(|e| Into::<Error>::into(e))
+                .and_then(|x| {
+                    tracing::debug!("Received new inclusion fee estimate {}...", x);
+
+                    let ret: Result<u64, _> = x
+                        .into_inner()
+                        .checked_div(FixedU128::accuracy())
+                        .ok_or(Error::ArithmeticUnderflow)
+                        .and_then(|x| x.try_into().map_err(Into::<Error>::into));
+                    futures::future::ready(ret)
+                })
+                .try_filter_map(|x| async move {
+                    match btc_rpc.fee_rate(txid) {
+                        Ok(current_fee) => {
+                            tracing::debug!("Current fee rate = {current_fee}. New estimate = {x}");
+                            if x > current_fee {
+                                Ok(Some((current_fee, x)))
+                            } else {
+                                Ok(None)
+                            }
+                        }
+                        Err(e) => {
+                            tracing::debug!("Failed to get fee_rate: {}", e);
+                            Ok(None)
+                        }
+                    }
+                })
+                .filter_map(|x| async {
+                    match btc_rpc.is_in_mempool(txid_copy) {
+                        Ok(false) => {
+                            //   if not in mempool anymore, don't propagate the event (even if it is an error)
+                            tracing::debug!("Txid not in mempool anymore...");
+                            None
+                        }
+                        Ok(true) => {
+                            tracing::debug!("Txid is still in mempool...");
+                            Some(x)
+                        }
+                        Err(e) => Some(Err(e.into())),
+                    }
+                });
+
+            let wait_for_transaction_metadata = btc_rpc.wait_for_transaction_metadata(txid, num_confirmations);
+            futures::pin_mut!(subscription);
+
+            use futures::future::Either;
+            let tx_metadata = match futures::future::select(wait_for_transaction_metadata, subscription.next()).await {
+                Either::Left((result, _)) => result?,
+                Either::Right((Some(Ok((old_fee, new_fee))), _)) => {
+                    tracing::info!("Bumping fee rate from {old_fee} to {new_fee}...");
+                    txid = self.bump_fee(btc_rpc, txid, new_fee).await?;
```

Comment:
Do we always want to bump the fee as soon as the estimate is higher that the current fee? If it's inexpensive or free I don't see why not but just want to confirm.

Reply:
Good point. Bumping fees does not cost the client anything, but there is a minimum increase for the network to accept the new tx. I believe that is set to 1 sat/vbyte, and since our price estimate has a granularity of 1 sat/vbyte, we could always bump. But in case 1 of these assumptions fails to hold (in the future) I will add an additional check for it

Reply:
I decided to just always try and catch the error